File: Yandex/Practicum/Algorightms_and_Data_Structures/1_sprint/H.Двоичная_система/solution.py
import sys


def solution(x: str, y: str) -> None:
    """
    Time Complexity O(N), where N is the length of place values of the largest number
    Memory Complexity O(N), where N is the length of place values of the largest number
    """
    digits: list[str] = []
    x_n: int = len(x)
    y_n: int = len(y)
    if x_n < y_n:
        x, y = y, x
        x_n, y_n = y_n, x_n
    y_idx: int = y_n - 1
    addition: int = 0
    for x_idx in range(x_n - 1, -1, -1):
        if y_idx >= 0:
            bits: int = int(x[x_idx]) + int(y[y_idx])
            if addition:
                bits += 1
                addition -= 1
            if bits > 1:
                digits.append('1' if bits == 3 else '0')
                addition += 1
            else:
                digits.append(str(bits))
            y_idx -= 1
        elif addition:
            bits: int = int(x[x_idx]) + addition
            addition -= 1
            if bits == 2:
                addition += 1
            digits.append('0' if bits == 2 else str(bits))
        else:
            digits.append(x[x_idx])

    if addition:
        digits.append('1')
    n: int = len(digits)
    for idx in range(n - 1, -1, -1):
        sys.stdout.write(digits[idx])
    sys.stdout.write('\n')


if __name__ == '__main__':
    x: str = sys.stdin.readline().rstrip()
    y: str = sys.stdin.readline().rstrip()
    solution(x, y)

File: Yandex/Practicum/Algorightms_and_Data_Structures/1_sprint/H.Двоичная_система/solution.cpp
#include <iostream>
#include <string>


void solution(std::string& x, std::string& y) {
    /*
    Time Complexity O(N), where N is the length of place values of the largest number
    Memory Complexity O(N), where N is the length of place values of the largest number
    */
    int xN = x.size(), yN = y.size();
    if (x.size() < y.size()) {
        std::string temp = x;
        int tempN = xN;
        x = y;
        xN = yN;
        y = temp;
        yN = tempN;
    }
    std::string digits = "";
    int yIdx = yN - 1;
    int additional = 0;
    for (int xIdx = xN - 1; xIdx >= 0; --xIdx) {
        if (yIdx >= 0) {
            int bits = (x[xIdx] == '0'? 0 : 1) + (y[yIdx] == '0'? 0 : 1);
            if (additional) {
                ++bits;
                --additional;
            }
            if (bits > 1) {
                ++additional;
                digits += (bits == 3? '1' : '0');
            } else {
                digits += std::to_string(bits);
            }
            --yIdx;
        } else if (additional) {
            int bits = (x[xIdx] == '0'? 0 : 1) + additional;
            --additional;
            if (bits == 2) {
                ++additional;
                digits += '0';
            } else {
                digits += std::to_string(bits);
            }
        } else {
            digits += x[xIdx];
        }
    }
    if (additional) {
        digits += '1';
    }
    int n = digits.size();
    for (int index = n - 1; index >= 0; --index) {
        std::cout << digits[index];
    }
    std::cout << "\n";
};


int main() {
    std::string x, y;
    std::cin >> x >> y;
    solution(x, y);
}
